client/my-sites/stats/hooks/use-moment-site-zone.ts

```diff
@@ -1,25 +1,29 @@
+import { createSelector } from '@automattic/state-utils';
 import i18n from 'i18n-calypso';
 import moment from 'moment';
 import { useSelector } from 'calypso/state';
 import { getSiteOption } from 'calypso/state/sites/selectors';
 import { getSelectedSiteId } from 'calypso/state/ui/selectors';
 
-export function getMomentSiteZone( state: object, siteId: number | null ) {
-	let localeSlug = i18n.getLocaleSlug();
-	if ( localeSlug === null ) {
-		localeSlug = 'en';
-	}
+export const getMomentSiteZone = createSelector(
+	( state: object, siteId: number | null ) => {
+		let localeSlug = i18n.getLocaleSlug();
+		if ( localeSlug === null ) {
+			localeSlug = 'en';
+		}
 
-	const localizedMoment = moment().locale( localeSlug );
+		const localizedMoment = moment().locale( localeSlug );
 
-	const gmtOffset = getSiteOption( state, siteId, 'gmt_offset' ) as number;
-	if ( Number.isFinite( gmtOffset ) ) {
-		return localizedMoment.utcOffset( gmtOffset );
-	}
+		const gmtOffset = getSiteOption( state, siteId, 'gmt_offset' ) as number;
+		if ( Number.isFinite( gmtOffset ) ) {
+			return localizedMoment.utcOffset( gmtOffset );
+		}
 
-	// Falls back to the browser's local timezone if no GMT offset is found
-	return localizedMoment;
-}
+		// Falls back to the browser's local timezone if no GMT offset is found
+		return localizedMoment;
+	},
+	[ ( state, siteId ) => getSiteOption( state, siteId, 'gmt_offset' ), () => i18n.getLocaleSlug() ]
+);
```

Comment:
Ideally this shouldn't be a Redux selector. The existing selector doesn't react to `i18n.getLocaleSlug()` changes. It should be a hook:
```js
function useMomentSiteZone() {
  const { localeSlug } = useTranslate()
  const siteOffset = useSelectedSiteGmtOffset();
  let m = moment().locale( localeSlug );
  if ( Number.isFinite( siteOffset ) {
    m = m.utcOffset( siteOffset );
  }
  return m;
}
```
Such a hook correctly reacts to all dependencies changing. And can return a new `moment` instance on each call without anyone caring.

Reply:
I don't like it to create new objects each time, but I feel like it's necessary because the moment object is mutable. In the usages, we try not to mutate the original object - clone before mutating, but it could be if not paying enough attention. With that regard, I think it'll still be better to return new object each time; however it does also depend on how expensive the operation is...

Reply:
Cloning `moment` objects should be extremely cheap. It's basically a shallow copy of a few fields. That's why it would be more ergonomic to just return a new object that you can freely mutate. Returning a memoized instance is dangerous, because if you mutate it, it has unexpected consequences.

Also, a rule of thumb for Redux is that the Redux state should store only "plain" JS values, not smart complex objects with prototypes and methods.